android/titanium/src/java/org/appcelerator/titanium/util/TiDeviceOrientation.java

```diff
@@ -0,0 +1,500 @@
+/**
+ * Appcelerator Titanium Mobile
+ * Copyright (c) 2009-2017 by Axway, Inc. All Rights Reserved.
+ * Licensed under the terms of the Apache Public License
+ * Please see the LICENSE included with this distribution for details.
+ */
+package org.appcelerator.titanium.util;
+
+import android.content.Context;
+import android.os.Build;
+import android.util.DisplayMetrics;
+import android.view.Display;
+import android.view.Surface;
+import android.view.WindowManager;
+import org.appcelerator.titanium.TiApplication;
+
+
+/**
+ * Enum type indicating if a device is portrait-upright, landscape-left, landscape-right, etc.
+ * <p>
+ * Provides methods to acquire its unique Titanium integer ID used in JavaScript
+ * and to determine if the device is portrait or landscape.
+ */
+public enum TiDeviceOrientation
+{
+	/** Indicates that the device orientation is unknown. */
+	UNKNOWN(TiIntegerId.UNKNOWN, "Ti.UI.UNKNOWN"),
```

Comment:
Can we simplify it a bit by using the enum as the sole definition and reducing `TiIntegerId` to just `id` (also make it `final`)

```Java
public enum TiDeviceOrientation
{
  UNKNOWN(0),
  PORTRAIT_UPRIGHT(1),
  PORTRAIT_UPSIDE_DOWN(2),
  LANDSCAPE_LEFT(3),
  LANDSCAPE_RIGHT(4),
  FACE_UP(5),
  FACE_DOWN(6);
  
  private final int id;
  
  private TiDeviceOrientation(int id)
  {
    this.id = id;
  }
  
  public int toInt()
  {
    return this.id;
  }

  ...
}
```

and then in our `UIModule.java` we can do this

```Java
@Kroll.constant public static final int PORTRAIT = TiDeviceOrientation.PORTRAIT_UPRIGHT.toInt();
@Kroll.constant public static final int UPSIDE_PORTRAIT = TiDeviceOrientation.PORTRAIT_UPSIDE_DOWN.toInt();
@Kroll.constant public static final int LANDSCAPE_LEFT = TiDeviceOrientation.LANDSCAPE_LEFT.toInt();
@Kroll.constant public static final int LANDSCAPE_RIGHT = TiDeviceOrientation.LANDSCAPE_RIGHT.toInt();
@Kroll.constant public static final int FACE_UP = TiDeviceOrientation.FACE_UP.toInt();
@Kroll.constant public static final int FACE_DOWN = TiDeviceOrientation.FACE_DOWN.toInt();
@Kroll.constant public static final int UNKNOWN = TiDeviceOrientation.UNKNOWN.toInt();
```

I'm not sure how I feel about storing the namespace as a string with the constant, since it means another thing to update. But if we do want to keep it we could do this instead:

```Java
private final String namespace = "Ti.UI." + this.name();
```

Reply:
Yes, we can set up our Kroll constant like this...
```java
@Kroll.constant public static final int PORTRAIT = TiDeviceOrientation.PORTRAIT_UPRIGHT.toTiIntegerId();
```

But the issue with this is that the Kroll constant can't be used in a switch statement. This is because the constant is assigned via a static initializer at runtime. The way I'm doing it, the constant is inlined at compile time and _**can**_ be used in a switch statement.

Reply:
We should not change `TiDeviceOrientation.toTiIntegerId()` to `toInt()`. The reason is to make the class future proof in case we want to provide other `toX()` style method mappings, such as to a Google Surface class rotation constant or an ActivityInfo screen orientation constant. I'm already doing this with the static `fromX()` methods I have now.

Also, I find that it's a good pattern to pair a `toX()` method and the static `fromX()` method with the same identifier name so that it's obvious to the developer that the 2 can be used in conjunction to serialize/deserialize back to the same object type.

```java
int id = TiDeviceOrientation.PORTRAIT_UPRIGHT.toTiIntegerId();
if (TiDeviceOrientation.fromTiIntegerId(id) == TiDeviceOrientation.PORTRAIT_UPRIGHT) {
	boolean isMatching = true;
}
```

Reply:
I just feel `toTiIntegerId()` is rather long. We could call it `toId()`/`fromId()` instead?